File: shortest_distance2.cpp
// Time Complexity : O(m+k) (Total number of matches found for word1: m, word2: n)
// Space Complexity : O(n)
// Any problem you faced while coding this :
// Your code here along with comments explaining your approach
//1. Create a hashmap to store all the words with occurence index
//2. Get the arrays from the hashmap and compare each index to find the min length using two pointers
//3. Return the min distance

class WordDistance {
public:
    unordered_map<string, vector<int>> word_map; 
    WordDistance(vector<string>& words) {
        int i=0;
        for(auto word:words){ //O(n) : but not counted because of the design
            word_map[word].push_back(i);
            i++;
        }
    }
    
    int shortest(string word1, string word2) {
        vector<int> vec1 = word_map[word1];
        vector<int> vec2 = word_map[word2];
        int v1=0, v2=0;
        int i1, i2;
        int min_dist= INT_MAX;
        while(v1<vec1.size() && v2<vec2.size()){
            i1=vec1[v1]; i2=vec2[v2];
            min_dist = min(min_dist, abs(i1-i2));
            if(i1>i2)
                v2++;
            else
                v1++;
        }
         return min_dist;
    }
};

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance* obj = new WordDistance(words);
 * int param_1 = obj->shortest(word1,word2);
 */

File: ShortestDistance2.java
//time complexity: O(n)
//space complexity: O(1)
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class WordDistance {
    HashMap<String,List<Integer>> map;
    public WordDistance(String[] wordsDict) {
        map=new HashMap<>();
        for(int i=0;i<wordsDict.length;i++)
        {
            if(map.get(wordsDict[i])==null)
                map.put(wordsDict[i],new ArrayList<Integer>());
            map.get(wordsDict[i]).add(i);
        }
    }
    
    public int shortest(String word1, String word2) {
        List<Integer> arr1=map.get(word1);
        List<Integer> arr2=map.get(word2);
        int p1=0;
        int p2=0;
        int max=Integer.MAX_VALUE;
        while(p1!=arr1.size() && p2!=arr2.size())
        {
            int ele1=arr1.get(p1);
            int ele2=arr2.get(p2);
            int diff=Math.abs(ele1-ele2);
            max=Math.min(diff,max);
            if(ele1>ele2)
                p2+=1;
            else
                p1+=1;
            
        }
        return max;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */